src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs

```diff
@@ -0,0 +1,372 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Linq;
+using System.Runtime.ConstrainedExecution;
+using System.Threading;
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.BackEnd.Components.Caching;
+using Microsoft.Build.BackEnd.Logging;
+using Microsoft.Build.BuildCheck.Acquisition;
+using Microsoft.Build.BuildCheck.Analyzers;
+using Microsoft.Build.BuildCheck.Logging;
+using Microsoft.Build.Collections;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Experimental.BuildCheck;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.BuildCheck.Infrastructure;
+
+internal delegate BuildAnalyzer BuildAnalyzerFactory();
+internal delegate BuildAnalyzerWrapper BuildAnalyzerWrapperFactory(ConfigurationContext configurationContext);
+
+/// <summary>
+/// The central manager for the BuildCheck - this is the integration point with MSBuild infrastructure.
+/// </summary>
+internal sealed class BuildCheckManagerProvider : IBuildCheckManagerProvider
+{
+    private static int s_isInitialized = 0;
+    private static IBuildCheckManager s_globalInstance = new NullBuildCheckManager();
+    internal static IBuildCheckManager GlobalInstance => s_isInitialized != 0 ? s_globalInstance : throw new InvalidOperationException("BuildCheckManagerProvider not initialized");
+
+    public IBuildCheckManager Instance => GlobalInstance;
+
+    internal static IBuildComponent CreateComponent(BuildComponentType type)
+    {
+        ErrorUtilities.VerifyThrow(type == BuildComponentType.BuildCheck, "Cannot create components of type {0}", type);
+        return new BuildCheckManagerProvider();
+    }
+
+    public void InitializeComponent(IBuildComponentHost host)
+    {
+        ErrorUtilities.VerifyThrow(host != null, "BuildComponentHost was null");
+
+        if (Interlocked.CompareExchange(ref s_isInitialized, 1, 0) == 1)
+        {
+            // Already initialized
+            return;
```

Comment:
This is not true. If `s_isInitialized` is 1, we know that another thread has made it this far. It hasn't necessarily initialized everything.

Reply:
I see this was fixed by a comment update. Is it really ok that initialization may not be finished when `InitializeComponent` returns?

I think it can be made safer and actually simpler by deleting `s_isInitialized` and synchronizing on `s_globalInstance` intself. I.e. (pseudo-code):

```
if (s_globalInstance is null)
{
  create new global instance
  InterlockedCompareExchange(try setting the new global instance into s_globalInstance if still null);
}
```

Reply:
Oh I see.
Yeah - I like this better - the construction is cheap, so we're fine with possible lost extra creation and the resulting code is shorter and more readable.

Thanks for clarifying! ;-)